packages/utils/src/useTableTools/useTableTools.js

```diff
@@ -0,0 +1,83 @@
+import { useEffect, useState } from 'react';
+import useFilterConfig from './useFilterConfig';
+import useTableSort from './useTableSort';
+import { default as usePaginate } from './usePaginate';
+import useRowsBuilder from './useRowsBuilder';
+import useBulkSelect from './useBulkSelect';
+
+export const useIdentify = (items, options = {}) => {
+    const getIdProp = (item, idProp = 'id') => (item[idProp]);
+    const identifier = options?.identifier || getIdProp;
+    const identify = (item) => {
+        if (typeof identifier === 'string') {
+            return getIdProp(item, identifier);
+        } else {
+            return identifier(item);
+        }
+    };
+
+    return {
+        items: items.map((item) => ({
+            ...item,
+            rowId: identify(item)
+        }))
+    };
+};
+
+const useTableTools = (items = [], columns = [], options) => {
+    const { page = 1, perPage = 10 } = options;
+    const [ tableState, setTableState ] = useState({
+        page,
+        perPage,
+        currentPageItemIds: []
+    });
+    const { items: identifiedItems } = useIdentify(items, options);
+    const {
+        items: selectedItems,
+        toolbarProps: bulkSelectToolbarProps,
+        tableProps: bulkSelectTableProps
+    } = options.selectable ? useBulkSelect(identifiedItems, options, [ tableState, setTableState ]) : { items: identifiedItems };
```

Comment:
A hook within a condition. Wouldn't this crash in some conditions, breaking the [Rules of Hooks](https://reactjs.org/docs/hooks-rules.html)?

Reply:
You are totally right, I forgot about that bit. I'm also not happy with the way the hooks are called here, since they somewhat depend on a certain order. I'll look over this again and try to refactor that part.